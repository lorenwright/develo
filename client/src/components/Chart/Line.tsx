import * as React from "react";
import {useRef, useEffect, useState} from "react";
import * as d3 from "d3";
import {ScaleLinear} from "d3";
import {Settings} from "../../types/ChartTypes";

type Props = {
  settings: Settings,
  data: object[],
  xScale: ScaleLinear<any, any>,
  yScale: ScaleLinear<any, any>,
  yAttribute: string,
  animate: boolean,
  chartWidth: number,
  showLabel: boolean
}

const Line: React.FC<Props> = ({settings, data, xScale, yScale, yAttribute, animate = false, chartWidth, showLabel}) => {
  const [labelDims, setLabelDims] = useState([])
  const path = useRef<SVGPathElement>(null)

  useEffect(() => {
    // The line we want to draw
    const line = d3.line()
      .x((d:any) => xScale(d[settings.xAttribute]))
      .y((d:any) => yScale(d[yAttribute]))
      .curve(d3.curveLinear)

    // Add the line to the chart
    d3.select(path.current)
      .datum(data)
      .attr('stroke', settings.lineStroke)
      .attr('stroke-width', settings.lineStrokeWidth)
      .attr('fill', 'none')
      .attr('d',
        // @ts-ignore
        line
      )

    // If showLabel is set, then we need to set the dimensions of where the label will be displayed
    if (data && data[data.length - 1] && showLabel) {
      setLabelDims([
        // @ts-ignore
        xScale(data[data.length - 1][settings.xAttribute] - 10), yScale(data[data.length - 1][yAttribute])
      ])
    }
  }, [data, chartWidth, settings, xScale, yScale, showLabel, yAttribute])

  return (
    <g>
      <path ref={path} className={`${animate ? 'draw-line' : ''}`} />
      {showLabel && labelDims.length > 0 && (
        <text transform={`translate(${labelDims[0]}, ${labelDims[1]})`} fill={settings.lineStroke} fontSize={10}>{yAttribute}</text>
      )}
    </g>
  )
}

export default Line;