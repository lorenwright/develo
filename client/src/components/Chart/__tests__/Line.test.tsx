import * as renderer from "react-test-renderer";
import Line from '../Line';
import '@testing-library/jest-dom';
import React from "react";
import * as d3 from "d3";

describe('Line', () => {
  it('renders correctly', () => {

    const defaultSettings = {
      height: 700,
      marginLeft: 30,
      marginRight: 0,
      marginTop: 5,
      marginBottom: 20,
      yMax: 100,
      yMin: 0,
      xMin: 0,
      xMax: 100,
      confidenceAreaFill: '#bae6fd',
      confidenceAreaStroke: 'none',
      lineStroke: '#38bdf8',
      lineStrokeWidth: '2',
      yTopAttribute: 'P97',
      yBottomAttribute: 'P3',
      xAttribute: 'Agemos',
      percentileLines: ['P3', 'P25', 'P50', 'P75', 'P97']
    }

    const range = [
      {P3: 0, P25: 1, P50: 2, P75: 3, P97: 4},
      {P3: 3, P25: 25, P50: 50, P75: 75, P97: 97},
    ]

    const chartWidth = 500

    const xScale = d3.scaleLinear()
        .domain([defaultSettings.xMin, defaultSettings.xMax])
        .range([defaultSettings.marginLeft, chartWidth - defaultSettings.marginLeft])

    const yScale = d3.scaleLinear()
        .domain([defaultSettings.yMax, defaultSettings.yMin])
        .range([defaultSettings.marginTop, defaultSettings.height - defaultSettings.marginBottom])

    const component = renderer.create(
        <Line settings={defaultSettings} data={range} xScale={xScale} yScale={yScale} yAttribute="P50" animate={false} chartWidth={chartWidth} showLabel />
    );

    let tree = component.toJSON();
    expect(tree).toMatchSnapshot();
  })
})