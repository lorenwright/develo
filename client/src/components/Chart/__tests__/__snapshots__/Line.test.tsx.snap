// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Line renders correctly 1`] = `
<g>
  <path
    className=""
  />
</g>
`;
